### ISSUE TITLE
Bad error message for `spack load` with no package specified

### ISSUE THREAD

**Issue by fengjingyu618** [spack/spack#10859]
Hi,

I enter the command at the terminal in the following order.

git clone https://github.com/spack/spack.git
. spack/share/spack/setup-env.sh
spack install boost+filesystem
spack install eigen
spack install tbb
spack install hdf5+hl
spack load

==> Error: the constraint '[]' matches multiple packages:
	/nrirv5h autoconf@2.69%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/ajob7oh automake@1.16.1%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/hg6c7ru boost@1.69.0+atomic+chrono~clanglibcpp cxxstd=default +date_time~debug+exception+fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199 ~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave%gcc@7.3.0+atomic+chrono~clanglibcpp cxxstd=default +date_time~debug+exception+fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199 ~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=linux-ubuntu18.04-x86_64 
	/4orv5th bzip2@1.0.6+shared%gcc@7.3.0+shared arch=linux-ubuntu18.04-x86_64 
	/sjpcmap cmake@3.13.4~doc+ncurses+openssl~ownlibs~qt%gcc@7.3.0~doc+ncurses+openssl~ownlibs~qt arch=linux-ubuntu18.04-x86_64 
	/dedz45f curl@7.63.0~darwinssl~libssh~libssh2~nghttp2%gcc@7.3.0~darwinssl~libssh~libssh2~nghttp2 arch=linux-ubuntu18.04-x86_64 
	/bskupck diffutils@3.7%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/w6xjeid eigen@3.3.5 build_type=RelWithDebInfo ~fftw~metis~mpfr~scotch~suitesparse%gcc@7.3.0 build_type=RelWithDebInfo ~fftw~metis~mpfr~scotch~suitesparse arch=linux-ubuntu18.04-x86_64 
	/zfnjp5i expat@2.2.5+libbsd%gcc@7.3.0+libbsd arch=linux-ubuntu18.04-x86_64 
	/cyayj4l gdbm@1.18.1%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/qtpajf6 gmp@6.1.2%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/uv32bt5 hdf5@1.10.4~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe%gcc@7.3.0~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu18.04-x86_64 
	/wuzckzw hwloc@1.11.11~cairo~cuda~gl+libxml2~nvml+pci+shared%gcc@7.3.0~cairo~cuda~gl+libxml2~nvml+pci+shared arch=linux-ubuntu18.04-x86_64 
	/3j7guip intel-tbb@2019.2 cxxstd=default patches=ca08c28bdb15582c30777f9303d1986e4c09b3d514776494f3fbf5f19381bfda +shared+tm%gcc@7.3.0 cxxstd=default patches=ca08c28bdb15582c30777f9303d1986e4c09b3d514776494f3fbf5f19381bfda +shared+tm arch=linux-ubuntu18.04-x86_64 
	/e6gfrng libarchive@3.3.2%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/urmxpfq libbsd@0.9.1%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/rblqbdg libelf@0.8.13%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/cun4ra6 libiconv@1.15%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/7x2dp3q libpciaccess@0.13.5%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/oromhtn libsigsegv@2.11%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/jher54m libtool@2.4.6%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/qpxxbs2 libuv@1.25.0%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/2whfhgo libxml2@2.9.8~python%gcc@7.3.0~python arch=linux-ubuntu18.04-x86_64 
	/inogpcg lz4@1.8.1.2%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/btessuo lzo@2.09%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/safarlc m4@1.4.18 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,c0a408fbffb7255fcc75e26bd8edab116fc81d216bfd18b473668b7739a4158e,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv%gcc@7.3.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,c0a408fbffb7255fcc75e26bd8edab116fc81d216bfd18b473668b7739a4158e,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-ubuntu18.04-x86_64 
	/c4iq7el ncurses@6.1~symlinks~termlib%gcc@7.3.0~symlinks~termlib arch=linux-ubuntu18.04-x86_64 
	/mkrgpv3 nettle@3.4%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/qo7ojsu numactl@2.0.12%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/bhtxi3r openmpi@3.1.3~cuda+cxx_exceptions fabrics=auto ~java~legacylaunchers~memchecker~pmi schedulers=auto ~sqlite3~thread_multiple+vt%gcc@7.3.0~cuda+cxx_exceptions fabrics=auto ~java~legacylaunchers~memchecker~pmi schedulers=auto ~sqlite3~thread_multiple+vt arch=linux-ubuntu18.04-x86_64 
	/jrtrw6d openssl@1.1.1b+systemcerts%gcc@7.3.0+systemcerts arch=linux-ubuntu18.04-x86_64 
	/7vsfnse perl@5.26.2+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads%gcc@7.3.0+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-ubuntu18.04-x86_64 
	/6cgd5ws pkgconf@1.5.4%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/lg4ce4z readline@7.0%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/fmamkjv rhash@1.3.5%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/325c2mb util-macros@1.19.1%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/buhe32j xz@5.2.4%gcc@7.3.0 arch=linux-ubuntu18.04-x86_64 
	/av3mmgf zlib@1.2.11+optimize+pic+shared%gcc@7.3.0+optimize+pic+shared arch=linux-ubuntu18.04-x86_64 

==> Error: In this context exactly **one** match is needed: please specify your constraints better.

Why it can't load installed dependencies?

Thanks in advance

**Comment by scheibelp** [comment]
`spack load` expects a spec that is specific enough to identify a single installed package. For example, assuming one copy of eigen is installed, `spack load eigen` would be sufficient.

We don't report the problem correctly so I added the "error-messages" label and I'll be updating the title of the bug.

Does that resolve the issue for you? If you are interested in loading groups of modules together for a set of installed packages, you can use an environment (see: https://spack.readthedocs.io/en/latest/tutorial_environments.html).

**Comment by fengjingyu618** [comment]
Thanks scheibelp.

yade2@ubuntu:~$ spack install eigen
==> eigen is already installed in /home/yade2/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/eigen-3.3.5-w6xjeid5rjfhjzgofpk7nuyrflngxv4r
yade2@ubuntu:~$ spack load eigen
module: command not found

Thanks,
Feng

**Comment by scheibelp** [comment]
It appears you don't have a module command available. You can install `environment-modules` with `spack bootstrap` or `spack install environment-modules` (and then add the `bin/` directory for that package to your `PATH`).

**Comment by alalazo** [comment]
Closing as this seems to be solved by #14062:
```console
$ spack find
==> 2 installed packages
-- linux-ubuntu18.04-broadwell / gcc@7.4.0 ----------------------
zlib@1.2.11

-- linux-ubuntu18.04-broadwell / gcc@9.0.1 ----------------------
zlib@1.2.11
$ spack load zlib
==> Error: zlib matches multiple packages.
  Matching packages:
    w4ulfwd zlib@1.2.11%gcc@7.4.0 arch=linux-ubuntu18.04-broadwell
    os3djb5 zlib@1.2.11%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
  Use a more specific spec.
$ spack load eigen
==> Error: Spec 'eigen' matches no installed packages.
```
@becker33
